librad/src/git/trailer.rs

```diff
@@ -0,0 +1,139 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::convert::TryFrom;
+
+#[derive(Debug, Clone, Eq, PartialEq, thiserror::Error)]
+pub enum Error {
+    #[error("the given input is not a well-formed trailer")]
+    IllFormed,
+
+    #[error("could not parse the trailer due to an invalid token found")]
+    Token(#[from] InvalidToken),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Trailer {
+    token: Token,
+    value: String,
+}
+
+impl Trailer {
+    #[allow(dead_code)]
+    /// Try and parse a [Trailer] form a given raw &str that is
+    /// expected to represent a trailer in raw form.
+    pub fn parse(s: &str, separators: &[char]) -> Result<Trailer, Error> {
+        let parts: Vec<String> = s
+            .split(&separators[..])
+            .map(|s| s.trim().to_string())
+            .collect();
+        match parts.as_slice() {
+            [t, v] => Ok(Trailer {
+                token: Token::try_from(t.clone())?,
+                value: v.clone(),
+            }),
+            _ => Err(Error::IllFormed),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Token(String);
+
+#[derive(Debug, Clone, Eq, PartialEq, thiserror::Error)]
+#[error("invalid token")]
+pub struct InvalidToken;
+
+impl TryFrom<String> for Token {
+    type Error = InvalidToken;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        // TODO(nuno): get this straight. The docs are not super clear.
+        // it has been pointed out that the Token should be an alphanumeric
+        // sequence of characters but that does not account for, at least, `-`,
+        // which is clearly supported.
+        if value.chars().all(|c| c.is_alphanumeric() || c == '-') {
+            Ok(Token(value))
+        } else {
+            Err(InvalidToken)
+        }
+    }
+}
+
+impl TryFrom<&str> for Token {
```

Comment:
`FromStr`?

Reply:
See https://github.com/rust-lang/api-guidelines/issues/182#issuecomment-479657836, keeping in mind that 2 years have passed and `TryFrom` is stable and recommended.

I'd be fine either way, but if we only _need_ one then I'd keep `TryFrom` unless if there's a good reason to prefer `FromStr`?

Reply:
We are using `FromStr` everywhere, not least because of `str::parse`. Feel free to _also_ provide `TryFrom`.

Reply:
Thanks. Moved to `FromStr` as default, using `TryFrom` as a convenience layer on top.